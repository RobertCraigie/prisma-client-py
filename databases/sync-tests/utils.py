from __future__ import annotations

import os
from pydantic import BaseModel

from .._types import DatabaseMapping
from .._compat import LiteralString


CURRENT_DATABASE = os.environ['PRISMA_DATABASE']


class RawQueries(BaseModel):
    """Raw queries defined globally so they can be easily referenced in test functions"""

    count_posts: LiteralString

    find_user_by_id: LiteralString
    find_user_by_id_limit_1: LiteralString

    find_post_by_id: LiteralString
    find_posts_not_published: LiteralString

    select_unknown_table: LiteralString
    update_unique_post_title: LiteralString
    update_unique_post_new_title: LiteralString
    test_query_raw_no_result: LiteralString
    test_execute_raw_no_result: LiteralString


_mysql_queries = RawQueries(
    count_posts="""
        SELECT COUNT(*) as count
        FROM Post
    """,
    find_post_by_id="""
        SELECT *
        FROM Post
        WHERE id = ?
    """,
    find_user_by_id="""
        SELECT *
        FROM User
        WHERE User.id = ?
    """,
    find_user_by_id_limit_1="""
        SELECT *
        FROM User
        WHERE User.id = ?
        LIMIT 1
    """,
    select_unknown_table="""
        SELECT *
        FROM bad_table;
    """,
    find_posts_not_published="""
        SELECT id, published
        FROM Post
        WHERE published = false
    """,
    test_query_raw_no_result="""
        SELECT *
        FROM Post
        WHERE id = 'sdldsd'
    """,
    update_unique_post_title="""
        UPDATE Post
        SET title = 'My edited title'
        WHERE id = ?
    """,
    update_unique_post_new_title="""
        UPDATE Post
        SET title = 'My new title'
        WHERE id = ?
    """,
    test_execute_raw_no_result="""
        UPDATE Post
        SET title = 'updated title'
        WHERE id = 'sdldsd'
    """,
)

_postgresql_queries = RawQueries(
    count_posts="""
        SELECT COUNT(*) as count
        FROM "Post"
    """,
    find_post_by_id="""
        SELECT *
        FROM "Post"
        WHERE id = $1
    """,
    find_user_by_id="""
        SELECT *
        FROM "User"
        WHERE "User".id = $1
    """,
    find_user_by_id_limit_1="""
        SELECT *
        FROM "User"
        WHERE "User".id = $1
        LIMIT 1
    """,
    select_unknown_table="""
        SELECT *
        FROM bad_table;
    """,
    find_posts_not_published="""
        SELECT id, published
        FROM "Post"
        WHERE published = false
    """,
    test_query_raw_no_result="""
        SELECT *
        FROM "Post"
        WHERE id = 'sdldsd'
    """,
    update_unique_post_title="""
        UPDATE "Post"
        SET title = 'My edited title'
        WHERE id = $1
    """,
    update_unique_post_new_title="""
        UPDATE "Post"
        SET title = 'My new title'
        WHERE id = $1
    """,
    test_execute_raw_no_result="""
        UPDATE "Post"
        SET title = 'updated title'
        WHERE id = 'sdldsd'
    """,
)

RAW_QUERIES_MAPPING: DatabaseMapping[RawQueries] = {
    'postgresql': _postgresql_queries,
    'cockroachdb': _postgresql_queries,
    'mysql': _mysql_queries,
    'mariadb': _mysql_queries,
    'sqlite': RawQueries(
        count_posts="""
            SELECT COUNT(*) as count
            FROM Post
        """,
        find_post_by_id="""
            SELECT *
            FROM Post
            WHERE id = ?
        """,
        find_user_by_id="""
            SELECT *
            FROM User
            WHERE User.id = ?
        """,
        find_user_by_id_limit_1="""
            SELECT *
            FROM User
            WHERE User.id = ?
            LIMIT 1
        """,
        select_unknown_table="""
            SELECT *
            FROM bad_table;
        """,
        find_posts_not_published="""
            SELECT id, published
            FROM Post
            WHERE published = false
        """,
        test_query_raw_no_result="""
            SELECT *
            FROM Post
            WHERE id = 'sdldsd'
        """,
        update_unique_post_title="""
            UPDATE Post
            SET title = 'My edited title'
            WHERE id = ?
        """,
        update_unique_post_new_title="""
            UPDATE Post
            SET title = 'My new title'
            WHERE id = ?
        """,
        test_execute_raw_no_result="""
            UPDATE Post
            SET title = 'updated title'
            WHERE id = 'sdldsd'
        """,
    ),
}
