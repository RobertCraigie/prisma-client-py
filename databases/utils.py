from __future__ import annotations

import os
from typing import Set
from pathlib import Path
from typing_extensions import Literal, get_args, override

from pydantic import BaseModel
from syrupy.location import PyTestLocation
from syrupy.extensions.amber import AmberSnapshotExtension

from ._types import DatabaseMapping
from ._compat import LiteralString

DatabaseFeature = Literal[
    'enum',
    'json',
    'date',
    'arrays',
    'array_push',
    'json_arrays',
    'raw_queries',
    'create_many_skip_duplicates',
    'transactions',
    'case_sensitivity',
    'full_text_search',
]


class DatabaseConfig(BaseModel):
    id: str
    name: str
    env_var: str
    bools_are_ints: bool
    autoincrement_id: str
    unsupported_features: Set[DatabaseFeature]
    default_date_func: str

    # TODO: run this under coverage
    def supports_feature(self, feature: DatabaseFeature) -> bool:  # pragma: no cover
        if feature not in get_args(DatabaseFeature):
            raise RuntimeError(f'Unknown feature: {feature}')

        return feature not in self.unsupported_features


# ------------------ Test helpers ------------------

from .constants import TESTS_DIR, SYNC_TESTS_DIR

SHARED_SNAPSHOTS_DIR = Path(__file__).parent.joinpath('__shared_snapshots__')
CURRENT_DATABASE = os.environ.get('PRISMA_DATABASE')


class AmberSharedExtension(AmberSnapshotExtension):
    """Syrupy extension that stores the snapshots in a parent __shared_snapshots__ dir"""

    @classmethod
    @override
    def dirname(cls, *, test_location: PyTestLocation) -> str:
        test_dir = Path(test_location.filepath).parent
        if test_dir.is_relative_to(SYNC_TESTS_DIR):
            rel_dir = test_dir.relative_to(SYNC_TESTS_DIR)
        else:
            rel_dir = test_dir.relative_to(TESTS_DIR)

        return str(SHARED_SNAPSHOTS_DIR.joinpath(rel_dir))


class RawQueries(BaseModel):
    """Raw queries defined globally so they can be easily referenced in test functions"""

    count_posts: LiteralString

    find_user_by_id: LiteralString
    find_user_by_id_limit_1: LiteralString

    find_post_by_id: LiteralString
    find_posts_not_published: LiteralString

    select_unknown_table: LiteralString
    update_unique_post_title: LiteralString
    update_unique_post_new_title: LiteralString
    test_query_raw_no_result: LiteralString
    test_execute_raw_no_result: LiteralString


_mysql_queries = RawQueries(
    count_posts="""
        SELECT COUNT(*) as count
        FROM Post
    """,
    find_post_by_id="""
        SELECT *
        FROM Post
        WHERE id = ?
    """,
    find_user_by_id="""
        SELECT *
        FROM User
        WHERE User.id = ?
    """,
    find_user_by_id_limit_1="""
        SELECT *
        FROM User
        WHERE User.id = ?
        LIMIT 1
    """,
    select_unknown_table="""
        SELECT *
        FROM bad_table;
    """,
    find_posts_not_published="""
        SELECT id, published
        FROM Post
        WHERE published = false
    """,
    test_query_raw_no_result="""
        SELECT *
        FROM Post
        WHERE id = 'sdldsd'
    """,
    update_unique_post_title="""
        UPDATE Post
        SET title = 'My edited title'
        WHERE id = ?
    """,
    update_unique_post_new_title="""
        UPDATE Post
        SET title = 'My new title'
        WHERE id = ?
    """,
    test_execute_raw_no_result="""
        UPDATE Post
        SET title = 'updated title'
        WHERE id = 'sdldsd'
    """,
)

_postgresql_queries = RawQueries(
    count_posts="""
        SELECT COUNT(*) as count
        FROM "Post"
    """,
    find_post_by_id="""
        SELECT *
        FROM "Post"
        WHERE id = $1
    """,
    find_user_by_id="""
        SELECT *
        FROM "User"
        WHERE "User".id = $1
    """,
    find_user_by_id_limit_1="""
        SELECT *
        FROM "User"
        WHERE "User".id = $1
        LIMIT 1
    """,
    select_unknown_table="""
        SELECT *
        FROM bad_table;
    """,
    find_posts_not_published="""
        SELECT id, published
        FROM "Post"
        WHERE published = false
    """,
    test_query_raw_no_result="""
        SELECT *
        FROM "Post"
        WHERE id = 'sdldsd'
    """,
    update_unique_post_title="""
        UPDATE "Post"
        SET title = 'My edited title'
        WHERE id = $1
    """,
    update_unique_post_new_title="""
        UPDATE "Post"
        SET title = 'My new title'
        WHERE id = $1
    """,
    test_execute_raw_no_result="""
        UPDATE "Post"
        SET title = 'updated title'
        WHERE id = 'sdldsd'
    """,
)

RAW_QUERIES_MAPPING: DatabaseMapping[RawQueries] = {
    'postgresql': _postgresql_queries,
    'cockroachdb': _postgresql_queries,
    'mysql': _mysql_queries,
    'mariadb': _mysql_queries,
    'sqlite': RawQueries(
        count_posts="""
            SELECT COUNT(*) as count
            FROM Post
        """,
        find_post_by_id="""
            SELECT *
            FROM Post
            WHERE id = ?
        """,
        find_user_by_id="""
            SELECT *
            FROM User
            WHERE User.id = ?
        """,
        find_user_by_id_limit_1="""
            SELECT *
            FROM User
            WHERE User.id = ?
            LIMIT 1
        """,
        select_unknown_table="""
            SELECT *
            FROM bad_table;
        """,
        find_posts_not_published="""
            SELECT id, published
            FROM Post
            WHERE published = false
        """,
        test_query_raw_no_result="""
            SELECT *
            FROM Post
            WHERE id = 'sdldsd'
        """,
        update_unique_post_title="""
            UPDATE Post
            SET title = 'My edited title'
            WHERE id = ?
        """,
        update_unique_post_new_title="""
            UPDATE Post
            SET title = 'My new title'
            WHERE id = ?
        """,
        test_execute_raw_no_result="""
            UPDATE Post
            SET title = 'updated title'
            WHERE id = 'sdldsd'
        """,
    ),
}
