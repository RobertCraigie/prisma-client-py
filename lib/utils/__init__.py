from ._utils import flatten as flatten
