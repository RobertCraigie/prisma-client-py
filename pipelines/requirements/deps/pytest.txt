pytest==7.4.4
