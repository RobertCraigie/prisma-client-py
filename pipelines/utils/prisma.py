from __future__ import annotations

import nox


def generate(
    session: nox.Session,
    *,
    schema: str | None = 'tests/data/schema.prisma',
    clean: bool = True,
) -> None:
    if clean:
        session.run('python', '-m', 'prisma_cleanup')

    if schema:
        args = (f'--schema={schema}',)
    else:
        args = ()

    session.run('prisma', 'generate', *args)
