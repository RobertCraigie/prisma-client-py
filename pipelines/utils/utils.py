from pathlib import Path

import nox


CACHE_DIR = Path.cwd() / '.cache'


def get_pkg_location(session: nox.Session, pkg: str) -> str:
    location = session.run(
        'python',
        '-c',
        f'import {pkg}; print({pkg}.__file__)',
        silent=True,
    )
    assert isinstance(location, str)
    return str(Path(location).parent)


def setup_coverage(session: nox.Session, identifier: str) -> None:
    if identifier:
        identifier = f'.{identifier}'

    coverage_file = f'.coverage.{session.name}{identifier}'
    session.env['COVERAGE_FILE'] = str(CACHE_DIR / coverage_file)


def setup_env(session: nox.Session) -> None:
    setup_coverage(
        session,
        identifier=session.python if isinstance(session.python, str) else '',
    )
    session.env['PRISMA_PY_DEBUG'] = '1'
    session.env['PYTEST_ADDOPTS'] = ' '.join(
        [
            f'"{opt}"'
            for opt in [
                '-W error',
                # httpx deprecation warnings
                "-W ignore:'cgi' is deprecated and slated for removal in Python 3.13:DeprecationWarning",
                '-W ignore:path is deprecated:DeprecationWarning',
            ]
        ]
    )
