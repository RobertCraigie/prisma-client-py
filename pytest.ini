[pytest]
markers =
  persist_data: Signifies that data should not be cleared between test runs.
  prisma: This marker must be used to enable model-based access.

# asyncio fixtures and tests must be explicitly marked
asyncio_mode = strict

# NOTE: this is required to stop pytest from loading conftest.py files
# for our integration tests, as pytest eagerly loads conftests backwards
# we cannot use a hook to ignore them as the integration conftest files
# are loaded before ours, we might be able to fix this by installing
# our own pytest plugin?
norecursedirs =
  integrations/*
  tests/integrations/postgresql/*
