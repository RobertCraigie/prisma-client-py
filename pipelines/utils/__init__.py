from ._lark import generate_parsers as generate_parsers
from .utils import *
