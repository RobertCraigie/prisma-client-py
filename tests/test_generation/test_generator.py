import sys
import subprocess
from pathlib import Path
from typing import Iterator
from distutils.dir_util import copy_tree

import pytest
from jinja2 import Environment, FileSystemLoader
from prisma.generator import BASE_PACKAGE_DIR, render_template, cleanup_templates
from prisma.generator.generator import OVERRIDING_TEMPLATES
from prisma.generator.utils import resolve_template_path

from ..utils import Testdir


def iter_templates_dir(path: Path) -> Iterator[Path]:
    templates = path / 'generator' / 'templates'
    assert templates.exists()

    for template in templates.iterdir():
        name = template.name

        if template.is_dir() or not name.endswith('.py.jinja') or name.startswith('_'):
            continue

        yield resolve_template_path(path, template.relative_to(templates))


def is_overrided_file(file: Path) -> bool:
    # NOTE: this will not work correctly if an overriding template
    # contains a path separator
    return file.name + '.jinja' in OVERRIDING_TEMPLATES


def assert_module_is_clean(path: Path) -> None:
    for template in iter_templates_dir(path):
        if is_overrided_file(template):
            if template.name == 'http.py':
                content = template.read_text()

                # basic check to ensure that the original file has been reinstated
                assert 'template' not in content
            else:  # pragma: no cover
                assert False, f'Unhandled check for {template}'
        else:
            assert not template.exists()


def assert_module_not_clean(path: Path) -> None:
    for template in iter_templates_dir(path):
        if is_overrided_file(template):
            if template.name == 'http.py':
                content = template.read_text()

                # basic check to ensure that the original file has been replaced
                assert 'template' in content
            else:  # pragma: no cover
                assert False, f'Unhandled check for {template}'
        else:
            assert template.exists()


def test_repeated_rstrip_bug(tmp_path: Path) -> None:
    """Previously, rendering schema.prisma.jinja would have rendered the file
    to schema.prism instead of schema.prisma
    """
    env = Environment(loader=FileSystemLoader(str(tmp_path)))

    template = 'schema.prisma.jinja'
    tmp_path.joinpath(template).write_text('foo')
    render_template(tmp_path, template, dict(), env=env)

    assert tmp_path.joinpath('schema.prisma').read_text() == 'foo'


def test_template_cleanup(testdir: Testdir) -> None:
    """Cleaning up templates removes all rendered files"""
    path = testdir.path / 'prisma'
    assert not path.exists()
    copy_tree(str(BASE_PACKAGE_DIR), str(path))

    assert_module_not_clean(path)
    cleanup_templates(path)
    assert_module_is_clean(path)

    # ensure cleaning an already clean module doesn't change anything
    cleanup_templates(path)
    assert_module_is_clean(path)


def test_template_cleanup_original_files_not_replaced(testdir: Testdir) -> None:
    """Generating the client twice does not override template backups"""
    path = testdir.path / 'prisma'
    assert not path.exists()

    # generate twice to ensure that originals of overriding templates
    # are not replaced
    testdir.generate()
    testdir.generate()

    assert_module_not_clean(path)
    cleanup_templates(path)
    assert_module_is_clean(path)


def test_erroneous_template_cleanup(testdir: Testdir) -> None:
    """Template runtime errors do not result in a partially generated module"""
    path = testdir.path / 'prisma'
    copy_tree(str(BASE_PACKAGE_DIR), str(path))

    assert_module_not_clean(path)

    template = '{{ undefined.foo }}'
    template_path = (
        testdir.path / 'prisma' / 'generator' / 'templates' / 'template.py.jinja'
    )
    template_path.write_text(template)

    with pytest.raises(subprocess.CalledProcessError) as exc:
        testdir.generate()

    output = str(exc.value.output, sys.getdefaultencoding())
    assert template in output

    assert_module_is_clean(path)
