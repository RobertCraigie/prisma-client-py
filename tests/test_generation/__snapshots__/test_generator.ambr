# name: test_error_handling
  list([
    '1 validation error for Config',
    'partial_type_generator.spec',
    "  Value error, Could not find a python file or module at foo [type=value_error, input_value='foo', input_type=str]",
  ])
# ---
